### PR TITLE
Improve instantiation of DependencyGraphBuilder

### PR DESCRIPTION
**Pull request by jvandort**
DependencyGraphBuilder was previously instantiated once for each resolution. Furthermore, it accepted a ResolutionStrategyInternal in its 'resolve' method.

The initial goal of this commit was to remove the resolution strategy as a parameter, as this is a public type, highly coupled to 'Configuration'. Now, we pass in only the relevant fields from the resolution strategy.

Given the number of parameters passed into the resolve method and constructor of the builder, it made sense to separate them based on services and configurable parameters. Now, we make the builder build scoped, and pass in all user-configurable values into the 'resolve' method.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ComponentResolversFactory.java [hunk 1]
```diff
@@ -62,68 +62,37 @@ public DefaultComponentResolversFactory(
...
-    ) {
-        return moduleDependencyResolverFactory.create(
+        resolvers.add(moduleDependencyResolverFactory.create(
             resolutionStrategy,
             repositories,
             metadataHandler.getComponentMetadataProcessorFactory(),
+            // `requestedAttributes` and `consumerSchema` are used to support filtering components by attributes
+            // when using dynamic versions. We should consider just removing that feature and making dynamic
+            // version selection dumber. Needing to know _what_ is being resolved (the `requestedAttributes`)
+            // should not be necessary to build a resolver that could theoretically resolve anything.
```
Comment (octylFractal) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ComponentResolversFactory.java:88]: 🎉 Thanks for the updated comment, I appreciate the extra detail.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultConfigurationResolver.java [hunk 2]
```diff
@@ -338,6 +336,56 @@ private static Path getConsumingProjectIdentityPath(ResolveContext resolveContex
...
+            }
+        }
+
+        // TODO: These dependencies should not be provided separately, but should be part of the root variant.
+        List<? extends DependencyMetadata> syntheticDependencies = includeSyntheticDependencies ?
+            resolveContext.getSyntheticDependencies() : Collections.emptyList();
+
+        // ResolveContext and ResolutionStrategy should not be passed into the dependency graph resolver.
+        // The resolver should be agnostic of these types so that it may be used to resolve
+        // non-Configuration types.
```
Comment (octylFractal) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultConfigurationResolver.java:371]: 💭 I'm confused about what this comment is for. These aren't passed in, are you worried people will try to refactor this to pass them in the future? If so, I would rather have this comment at the declaration site of `DependencyGraphResolver#resolve`.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java [hunk 3]
```diff
@@ -41,10 +41,14 @@
 import java.util.Set;
 
 public class DefaultCapabilitiesConflictHandler implements CapabilitiesConflictHandler {
-    private final List<Resolver> resolvers = Lists.newArrayListWithExpectedSize(3);
+    private final List<Resolver> resolvers;
```
Comment (octylFractal) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/conflicts/DefaultCapabilitiesConflictHandler.java:44]: 🎉 Glad to see this become fully immutable.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultConfigurationResolver.java (hunk 2 above)
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultConfigurationResolver.java:371]: You are correct. 

Moving this to `DependencyGraphResolver#resolve`

**Review by octylFractal — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Review by big-guy — APPROVED** [review]

**Pull request merged**